pallets/collective-proxy/src/lib.rs

```diff
@@ -121,5 +178,66 @@ pub mod pallet {
 
             Ok(())
         }
+
+        /// Register a proxy account for the sender that is able to make calls on its behalf.
+        ///
+        /// The dispatch origin for this call must be _Signed_.
+        ///
+        /// Parameters:
+        /// - `proxy`: The account that the `caller` would like to make a proxy.
+        /// - `filter`: Call filter used for the proxy
+        #[pallet::call_index(1)]
+        #[pallet::weight(T::WeightInfo::add_proxy(T::MaxProxies::get()))]
+        pub fn add_proxy(
+            origin: OriginFor<T>,
+            proxy: T::AccountId,
+            filter: T::CallFilter,
+        ) -> DispatchResult {
+            T::CollectiveProxy::ensure_origin(origin)?;
+            Proxies::<T>::try_mutate(|proxies| -> Result<(), DispatchError> {
+                let proxy_def = ProxyDefinition {
+                    proxy: proxy.clone(),
+                    filter: filter.clone(),
+                };
+                proxies.try_push(proxy_def).map_err(|_| Error::<T>::TooManyProxies)?;
+                Ok(())
+            })
+        }
```

Comment:
What happens if there are multiple _Proxies_ for the same account Id?

Reply:
They differ by calling filter and the first found with the suitable filter will be used. I think, I relied this logic on the base proxy pallet

Reply:
Aa, I got your concern. I think, it can be easily addressed with superset validation that would prevent the case you're describing (the change added).

Reply:
I see!

Out of curiosity, how much less performant was the `BTreeMap` (I'm assuming you tried using that 🙂 )?

Back to my suggestion above, performance wise, this should be equivalent or faster than the vec approach:
```Rust
    #[pallet::storage]
    pub type Proxies<T: Config> = StorageMap<
        _,
        (T::CollectiveProxy, T::AccountId),
        T::CallFilter,
        ValueQuery,
    >;
```

The `T::CollectiveProxy` would be extended to be a parameter-like type.
The `T::CallFilter` would follow your approach.

This way you can support more than 1 collective proxy type, each proxy can delegate to multiple accounts, but only one delegation type is possible.

Anyways, just something of top of my hat, haven't tried or implemented it 🙂 

---

Thank you for all the replies & the effort.
Except for the one at the start of this comment, no more questions from me :)

Reply:
I was considering standard BoundedBTreeMap. I don't think, there is any significant difference in performance for such small sets of data (adding to equation codec's complexity makes things even more complicated). My concern was mostly about the additional storage costs for TreeMap. I don't remember already but there was a general optimization hint to use compact storage structures as much as possible. 

Thank you for your comments and questions! It helped me a lot to understand better the use case.